## Technique 6-10

* 사용 데이터: 피자 체인 데이터

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn
import os
import glob

In [7]:
# 데이터 로드
os.chdir('/Users/yun-yeongjun/Desktop/Technique100/PythonML-main/PythonML-main/chapter01')

m_store = pd.read_csv('m_store.csv')
m_area = pd.read_csv('m_area.csv')

order_all = pd.DataFrame()
tbl_order_file = os.path.join(os.getcwd(), 'tbl_order_*.csv')
tbl_order_files = glob.glob(tbl_order_file)

for file in tbl_order_files:
    order_data = pd.read_csv(file)
    print(f'{file}:{len(order_data)}')
    order_all = pd.concat([order_all, order_data], ignore_index=True)
    
order_all

/Users/yun-yeongjun/Desktop/Technique100/PythonML-main/PythonML-main/chapter01/tbl_order_202104.csv:233262
/Users/yun-yeongjun/Desktop/Technique100/PythonML-main/PythonML-main/chapter01/tbl_order_202105.csv:241142
/Users/yun-yeongjun/Desktop/Technique100/PythonML-main/PythonML-main/chapter01/tbl_order_202106.csv:233302


order_id  store_id customer_id  coupon_cd  sales_detail_id  \
0       34104383        11   C65806632         57         61573513   
1       70652318        59   C09760173         37         54068709   
2       71640388       195   C61227084         17         93678366   
3       75673365       127   C64119972         17          5287952   
4        9077529       174   C10231192         18         18248867   
...          ...       ...         ...        ...              ...   
707701  17154702       157   C83348111         25         15969851   
707702  49731888       185   C55784926         55         96862999   
707703  93805252        79   C87712110         43         50807591   
707704  33290611       116   C83096347         59         91424716   
707705  84666999        97   C55599659         94         99820440   

       order_accept_date    delivered_date  takeout_flag  total_amount  status  
0       2021-04-01 11:00  2021-04-01 11:39             1         28270       1  
1       2021-04-01 11:00  2021-04-01 11:34             0         28270       2  
2       2021-04-01 11:00  2021-04-01 11:54             0         26470       9  
3       2021-04-01 11:00  2021-04-01 11:17             0         23080       2  
4       2021-04-01 11:00  2021-04-01 11:35             0         46920       2  
...                  ...               ...           ...           ...     ...  
707701  2021-06-30 21:58  2021-06-30 22:35             0         44620       2  
707702  2021-06-30 21:58  2021-06-30 22:34             1         23280       1  
707703  2021-06-30 21:58  2021-06-30 22:12             0         26150       2  
707704  2021-06-30 21:58  2021-06-30 22:31             1         35020       1  
707705  2021-06-30 21:58  2021-06-30 22:41             0         23120       2  

[707706 rows x 10 columns]

#### 6. 불필요한 데이터를 제거하자

In [11]:
# m_store를 다시 보면 store_id에 보수담당으로 999가 있음 -> 불필요
# store_id = 999는 유지보수용 데이터로, order 테이블에도 있을 것으로 생각된다.
m_store

store_id store_name area_cd
0           1      삼일대로점      SL
1           2      세종대로점      SL
2           3       무교로점      SL
3           4      덕수궁길점      SL
4           5      서소문로점      SL
..        ...        ...     ...
192       193       방죽로점      GB
193       194      일산동구점      GB
194       195       화중로점      GB
195       196       석봉로점      GB
196       999       보수담당      SL

[197 rows x 3 columns]

In [9]:
order_all

order_id  store_id customer_id  coupon_cd  sales_detail_id  \
0       34104383        11   C65806632         57         61573513   
1       70652318        59   C09760173         37         54068709   
2       71640388       195   C61227084         17         93678366   
3       75673365       127   C64119972         17          5287952   
4        9077529       174   C10231192         18         18248867   
...          ...       ...         ...        ...              ...   
707701  17154702       157   C83348111         25         15969851   
707702  49731888       185   C55784926         55         96862999   
707703  93805252        79   C87712110         43         50807591   
707704  33290611       116   C83096347         59         91424716   
707705  84666999        97   C55599659         94         99820440   

       order_accept_date    delivered_date  takeout_flag  total_amount  status  
0       2021-04-01 11:00  2021-04-01 11:39             1         28270       1  
1       2021-04-01 11:00  2021-04-01 11:34             0         28270       2  
2       2021-04-01 11:00  2021-04-01 11:54             0         26470       9  
3       2021-04-01 11:00  2021-04-01 11:17             0         23080       2  
4       2021-04-01 11:00  2021-04-01 11:35             0         46920       2  
...                  ...               ...           ...           ...     ...  
707701  2021-06-30 21:58  2021-06-30 22:35             0         44620       2  
707702  2021-06-30 21:58  2021-06-30 22:34             1         23280       1  
707703  2021-06-30 21:58  2021-06-30 22:12             0         26150       2  
707704  2021-06-30 21:58  2021-06-30 22:31             1         35020       1  
707705  2021-06-30 21:58  2021-06-30 22:41             0         23120       2  

[707706 rows x 10 columns]

In [10]:
order_all = order_all.loc[order_all['store_id'] != 999]
order_all

order_id  store_id customer_id  coupon_cd  sales_detail_id  \
0       34104383        11   C65806632         57         61573513   
1       70652318        59   C09760173         37         54068709   
2       71640388       195   C61227084         17         93678366   
3       75673365       127   C64119972         17          5287952   
4        9077529       174   C10231192         18         18248867   
...          ...       ...         ...        ...              ...   
707701  17154702       157   C83348111         25         15969851   
707702  49731888       185   C55784926         55         96862999   
707703  93805252        79   C87712110         43         50807591   
707704  33290611       116   C83096347         59         91424716   
707705  84666999        97   C55599659         94         99820440   

       order_accept_date    delivered_date  takeout_flag  total_amount  status  
0       2021-04-01 11:00  2021-04-01 11:39             1         28270       1  
1       2021-04-01 11:00  2021-04-01 11:34             0         28270       2  
2       2021-04-01 11:00  2021-04-01 11:54             0         26470       9  
3       2021-04-01 11:00  2021-04-01 11:17             0         23080       2  
4       2021-04-01 11:00  2021-04-01 11:35             0         46920       2  
...                  ...               ...           ...           ...     ...  
707701  2021-06-30 21:58  2021-06-30 22:35             0         44620       2  
707702  2021-06-30 21:58  2021-06-30 22:34             1         23280       1  
707703  2021-06-30 21:58  2021-06-30 22:12             0         26150       2  
707704  2021-06-30 21:58  2021-06-30 22:31             1         35020       1  
707705  2021-06-30 21:58  2021-06-30 22:41             0         23120       2  

[703886 rows x 10 columns]

* store_id = 999인 데이터를 제거하니 데이터가 707706개에서 703886개로 줄어들었다.
    * 이런 유지보수용 데이터는 언제 만들어졌는지 알 수 없기 때문에 미리 제거하는 것이 좋다.
   
***
* loc: location의 약어
    * df의 행이나 컬럼에 label이나 boolean array로 접근
    * 인간이 읽을 수 있는 label값으로 데이터에 접근한다.
    * 헷갈리는 메서드로 iloc가 있음
* iloc: integer location의 약어
    * df의 행이나 컬럼에 index값으로 접근
    * 컴퓨터가 읽을 수 있는 indexing값으로 데이터에 접근한다.

In [12]:
# 데이터가 줄어들었으니 통계량에도 변화가 생겼을 것이다.
order_all.describe()

order_id       store_id      coupon_cd  sales_detail_id  \
count  7.038860e+05  703886.000000  703886.000000     7.038860e+05   
mean   5.001648e+07      99.077635      49.514404     4.999685e+07   
std    2.889610e+07      55.942150      28.883529     2.885701e+07   
min    2.200000e+01       1.000000       0.000000     4.600000e+01   
25%    2.501640e+07      51.000000      25.000000     2.503448e+07   
50%    5.000843e+07      99.000000      49.000000     5.003433e+07   
75%    7.509089e+07     147.000000      75.000000     7.499115e+07   
max    9.999972e+07     196.000000      99.000000     9.999964e+07   

        takeout_flag   total_amount        status  
count  703886.000000  703886.000000  703886.00000  
mean        0.259897   29599.481351       3.08279  
std         0.438578    9543.594711       2.83614  
min         0.000000    6980.000000       1.00000  
25%         0.000000   23080.000000       2.00000  
50%         0.000000   28080.000000       2.00000  
75%         1.000000   36170.000000       2.00000  
max         1.000000   51000.000000       9.00000

#### 마스터 데이터를 조인하자.

In [16]:
# 마스터 데이터를 가로로 조인
# 매장별 주문 건수를 확인하기 위해 m_store테이블을 order_all에 조인
# 각 데이터의 공통열은 'store_id'
# how= 로 주축이 될 테이블을 정한다.

order_data = pd.merge(order_all, m_store, on='store_id', how='left')
order_data

order_id  store_id customer_id  coupon_cd  sales_detail_id  \
0       34104383        11   C65806632         57         61573513   
1       70652318        59   C09760173         37         54068709   
2       71640388       195   C61227084         17         93678366   
3       75673365       127   C64119972         17          5287952   
4        9077529       174   C10231192         18         18248867   
...          ...       ...         ...        ...              ...   
703881  17154702       157   C83348111         25         15969851   
703882  49731888       185   C55784926         55         96862999   
703883  93805252        79   C87712110         43         50807591   
703884  33290611       116   C83096347         59         91424716   
703885  84666999        97   C55599659         94         99820440   

       order_accept_date    delivered_date  takeout_flag  total_amount  \
0       2021-04-01 11:00  2021-04-01 11:39             1         28270   
1       2021-04-01 11:00  2021-04-01 11:34             0         28270   
2       2021-04-01 11:00  2021-04-01 11:54             0         26470   
3       2021-04-01 11:00  2021-04-01 11:17             0         23080   
4       2021-04-01 11:00  2021-04-01 11:35             0         46920   
...                  ...               ...           ...           ...   
703881  2021-06-30 21:58  2021-06-30 22:35             0         44620   
703882  2021-06-30 21:58  2021-06-30 22:34             1         23280   
703883  2021-06-30 21:58  2021-06-30 22:12             0         26150   
703884  2021-06-30 21:58  2021-06-30 22:31             1         35020   
703885  2021-06-30 21:58  2021-06-30 22:41             0         23120   

        status store_name area_cd  
0            1       자양로점      SL  
1            2     가마산로2점      SL  
2            9       화중로점      GB  
3            2       분포로점      BS  
4            2      하남산단점      GJ  
...        ...        ...     ...  
703881       2       우치로점      GJ  
703882       1       조치원점      SJ  
703883       2       종로2점      SL  
703884       1   해안새벽시장길점      BS  
703885       2       연제로점      BS  

[703886 rows x 12 columns]

In [18]:
# order_data에 지역 마스터를 조인

order_data = pd.merge(order_data, m_area, on='area_cd', how='left')
order_data

order_id  store_id customer_id  coupon_cd  sales_detail_id  \
0       34104383        11   C65806632         57         61573513   
1       70652318        59   C09760173         37         54068709   
2       71640388       195   C61227084         17         93678366   
3       75673365       127   C64119972         17          5287952   
4        9077529       174   C10231192         18         18248867   
...          ...       ...         ...        ...              ...   
703881  17154702       157   C83348111         25         15969851   
703882  49731888       185   C55784926         55         96862999   
703883  93805252        79   C87712110         43         50807591   
703884  33290611       116   C83096347         59         91424716   
703885  84666999        97   C55599659         94         99820440   

       order_accept_date    delivered_date  takeout_flag  total_amount  \
0       2021-04-01 11:00  2021-04-01 11:39             1         28270   
1       2021-04-01 11:00  2021-04-01 11:34             0         28270   
2       2021-04-01 11:00  2021-04-01 11:54             0         26470   
3       2021-04-01 11:00  2021-04-01 11:17             0         23080   
4       2021-04-01 11:00  2021-04-01 11:35             0         46920   
...                  ...               ...           ...           ...   
703881  2021-06-30 21:58  2021-06-30 22:35             0         44620   
703882  2021-06-30 21:58  2021-06-30 22:34             1         23280   
703883  2021-06-30 21:58  2021-06-30 22:12             0         26150   
703884  2021-06-30 21:58  2021-06-30 22:31             1         35020   
703885  2021-06-30 21:58  2021-06-30 22:41             0         23120   

        status store_name area_cd wide_area narrow_area  
0            1       자양로점      SL        서울          서울  
1            2     가마산로2점      SL        서울          서울  
2            9       화중로점      GB        경기        경기북부  
3            2       분포로점      BS        부산          부산  
4            2      하남산단점      GJ        광주          광주  
...        ...        ...     ...       ...         ...  
703881       2       우치로점      GJ        광주          광주  
703882       1       조치원점      SJ        세종          세종  
703883       2       종로2점      SL        서울          서울  
703884       1   해안새벽시장길점      BS        부산          부산  
703885       2       연제로점      BS        부산          부산  

[703886 rows x 14 columns]

#### 마스터가 존재하지 않는 코드에 이름을 설정하자

* order_data에서 takeout_flag, status 컬럼에 값은 있지만, 이 값이 무엇을 의미하는지 테이블만 봐서는 알 수 없다.
    * 책에서는 takeout_flag 컬럼의 값 중, 0= delivery, 1= takeout이라고 한다.
    * 또한 status 컬럼의 값 중, 0: 주문 접수, 1: 지불 완료, 2: 배달 완료, 9: 주문 취소

In [19]:
# order_data에 takeout_name 컬럼을 추가하고 각각의 값에 연결

order_data.loc[order_data['takeout_flag'] == 0, 'takeout_name'] = 'delivery'
order_data.loc[order_data['takeout_flag'] == 1, 'takeout_name'] = 'takeout'
order_data

order_id  store_id customer_id  coupon_cd  sales_detail_id  \
0       34104383        11   C65806632         57         61573513   
1       70652318        59   C09760173         37         54068709   
2       71640388       195   C61227084         17         93678366   
3       75673365       127   C64119972         17          5287952   
4        9077529       174   C10231192         18         18248867   
...          ...       ...         ...        ...              ...   
703881  17154702       157   C83348111         25         15969851   
703882  49731888       185   C55784926         55         96862999   
703883  93805252        79   C87712110         43         50807591   
703884  33290611       116   C83096347         59         91424716   
703885  84666999        97   C55599659         94         99820440   

       order_accept_date    delivered_date  takeout_flag  total_amount  \
0       2021-04-01 11:00  2021-04-01 11:39             1         28270   
1       2021-04-01 11:00  2021-04-01 11:34             0         28270   
2       2021-04-01 11:00  2021-04-01 11:54             0         26470   
3       2021-04-01 11:00  2021-04-01 11:17             0         23080   
4       2021-04-01 11:00  2021-04-01 11:35             0         46920   
...                  ...               ...           ...           ...   
703881  2021-06-30 21:58  2021-06-30 22:35             0         44620   
703882  2021-06-30 21:58  2021-06-30 22:34             1         23280   
703883  2021-06-30 21:58  2021-06-30 22:12             0         26150   
703884  2021-06-30 21:58  2021-06-30 22:31             1         35020   
703885  2021-06-30 21:58  2021-06-30 22:41             0         23120   

        status store_name area_cd wide_area narrow_area takeout_name  
0            1       자양로점      SL        서울          서울      takeout  
1            2     가마산로2점      SL        서울          서울     delivery  
2            9       화중로점      GB        경기        경기북부     delivery  
3            2       분포로점      BS        부산          부산     delivery  
4            2      하남산단점      GJ        광주          광주     delivery  
...        ...        ...     ...       ...         ...          ...  
703881       2       우치로점      GJ        광주          광주     delivery  
703882       1       조치원점      SJ        세종          세종      takeout  
703883       2       종로2점      SL        서울          서울     delivery  
703884       1   해안새벽시장길점      BS        부산          부산      takeout  
703885       2       연제로점      BS        부산          부산     delivery  

[703886 rows x 15 columns]

In [20]:
# status도 위처럼 추가

order_data.loc[order_data['status'] == 0, 'status_name'] = '주문 접수'
order_data.loc[order_data['status'] == 1, 'status_name'] = '지불 완료'
order_data.loc[order_data['status'] == 2, 'status_name'] = '배달 완료'
order_data.loc[order_data['status'] == 9, 'status_name'] = '주문 취소'

order_data

order_id  store_id customer_id  coupon_cd  sales_detail_id  \
0       34104383        11   C65806632         57         61573513   
1       70652318        59   C09760173         37         54068709   
2       71640388       195   C61227084         17         93678366   
3       75673365       127   C64119972         17          5287952   
4        9077529       174   C10231192         18         18248867   
...          ...       ...         ...        ...              ...   
703881  17154702       157   C83348111         25         15969851   
703882  49731888       185   C55784926         55         96862999   
703883  93805252        79   C87712110         43         50807591   
703884  33290611       116   C83096347         59         91424716   
703885  84666999        97   C55599659         94         99820440   

       order_accept_date    delivered_date  takeout_flag  total_amount  \
0       2021-04-01 11:00  2021-04-01 11:39             1         28270   
1       2021-04-01 11:00  2021-04-01 11:34             0         28270   
2       2021-04-01 11:00  2021-04-01 11:54             0         26470   
3       2021-04-01 11:00  2021-04-01 11:17             0         23080   
4       2021-04-01 11:00  2021-04-01 11:35             0         46920   
...                  ...               ...           ...           ...   
703881  2021-06-30 21:58  2021-06-30 22:35             0         44620   
703882  2021-06-30 21:58  2021-06-30 22:34             1         23280   
703883  2021-06-30 21:58  2021-06-30 22:12             0         26150   
703884  2021-06-30 21:58  2021-06-30 22:31             1         35020   
703885  2021-06-30 21:58  2021-06-30 22:41             0         23120   

        status store_name area_cd wide_area narrow_area takeout_name  \
0            1       자양로점      SL        서울          서울      takeout   
1            2     가마산로2점      SL        서울          서울     delivery   
2            9       화중로점      GB        경기        경기북부     delivery   
3            2       분포로점      BS        부산          부산     delivery   
4            2      하남산단점      GJ        광주          광주     delivery   
...        ...        ...     ...       ...         ...          ...   
703881       2       우치로점      GJ        광주          광주     delivery   
703882       1       조치원점      SJ        세종          세종      takeout   
703883       2       종로2점      SL        서울          서울     delivery   
703884       1   해안새벽시장길점      BS        부산          부산      takeout   
703885       2       연제로점      BS        부산          부산     delivery   

       status_name  
0            지불 완료  
1            배달 완료  
2            주문 취소  
3            배달 완료  
4            배달 완료  
...            ...  
703881       배달 완료  
703882       지불 완료  
703883       배달 완료  
703884       지불 완료  
703885       배달 완료  

[703886 rows x 16 columns]

#### 분석 기초 테이블을 파일에 저장하자.

In [22]:
# 나는 os.chdir로 경로를 애초에 설정해두거나 따로 경로를 저장한 변수로 바로 to_csv 저장하는 것을 선호
# 책에서는(24p) 경로 상에 폴더를 생성하는 os.makedir 메서드를 사용하여 저장하는 방식, 이게 더 편한 것 같기도..

store_path = '/Users/yun-yeongjun/Desktop/Technique100/make_data'
store_file = os.path.join(store_path, 'order_data.csv')
order_data.to_csv(store_file, index=False)

#### 셀을 사용하기 쉽게 정리하자

1. 처리 단위를 고려해 셀의 수를 줄이는 것이 좋다.
2. import를 앞쪽으로 모은다.
3. 적절한 주석 추가
4. 데이터 확인용 코드는 반드시 필요한 것이 아니면 제거